community/bolt/src/main/java/org/neo4j/bolt/v1/transport/ChunkedOutput.java

```diff
@@ -24,52 +24,92 @@
 
 import java.io.IOException;
 import java.nio.ByteBuffer;
-import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.Objects;
 
 import org.neo4j.bolt.transport.TransportThrottleException;
 import org.neo4j.bolt.transport.TransportThrottleGroup;
 import org.neo4j.bolt.v1.messaging.BoltIOException;
-import org.neo4j.bolt.v1.messaging.BoltResponseMessageBoundaryHook;
 import org.neo4j.bolt.v1.packstream.PackOutput;
 import org.neo4j.bolt.v1.packstream.PackOutputClosedException;
 import org.neo4j.bolt.v1.packstream.PackStream;
 import org.neo4j.kernel.api.exceptions.Status;
 
-import static java.lang.Math.max;
-
 /**
  * A target output for {@link PackStream} which breaks the data into a continuous stream of chunks before pushing them into a netty
  * channel.
  */
-public class ChunkedOutput implements PackOutput, BoltResponseMessageBoundaryHook
+public class ChunkedOutput implements PackOutput
 {
     public static final int CHUNK_HEADER_SIZE = 2;
     public static final int MESSAGE_BOUNDARY = 0;
 
-    private final int bufferSize;
+    private static final int MAX_CHUNK_SIZE = Short.MAX_VALUE / 2;
+    private static final int NO_MESSAGE = -1;
+
+    private final Channel channel;
+    private final int maxBufferSize;
     private final int maxChunkSize;
-    private final AtomicBoolean closed = new AtomicBoolean( false );
     private final TransportThrottleGroup throttleGroup;
-    private final Channel channel;
 
     private ByteBuf buffer;
-    private int currentChunkHeaderOffset;
+    private int currentChunkStartIndex;
+    private boolean closed;
 
     /** Are currently in the middle of writing a chunk? */
     private boolean chunkOpen;
+    private int currentMessageStartIndex = NO_MESSAGE;
 
-    public ChunkedOutput( Channel ch, int bufferSize )
+    public ChunkedOutput( Channel ch, int bufferSize, TransportThrottleGroup throttleGroup )
     {
-        this( ch, bufferSize, null );
+        this( ch, bufferSize, MAX_CHUNK_SIZE, throttleGroup );
     }
 
-    public ChunkedOutput( Channel ch, int bufferSize, TransportThrottleGroup throttleGroup )
+    public ChunkedOutput( Channel channel, int maxBufferSize, int maxChunkSize, TransportThrottleGroup throttleGroup )
+    {
+        this.channel = Objects.requireNonNull( channel );
+        this.maxBufferSize = maxBufferSize;
+        this.maxChunkSize = maxChunkSize;
+        this.buffer = allocateBuffer();
+        this.throttleGroup = Objects.requireNonNull( throttleGroup );
+    }
+
+    @Override
+    public synchronized void beginMessage()
+    {
+        if ( currentMessageStartIndex != NO_MESSAGE )
+        {
+            throw new IllegalStateException( "Message has already been started, index: " + currentMessageStartIndex );
+        }
+
+        currentMessageStartIndex = buffer.writerIndex();
```

Comment:
Would it be better if we used `ByteBuf#markWriterIndex()` and `ByteBuf#resetWriterIndex()` for writer index tracking?

Reply:
Did not know such methods exist :)
Looks like there is no way to discard a marked index or set it to some invalid number, like -1. That is why I think current approach is safer because validation like in `#assertMessageStarted()` is possible.